src/Lecture3.hs

```diff
@@ -246,4 +330,26 @@ Just [8,9,10]
 [8,20,3]
 
 -}
-apply = error "TODO"
+apply :: Functor t => a -> t (a -> b) -> t b
+apply e f = fmap ($ e) f 
+-- this took me many hours to figure out...and I don't understand why it works
+-- since the first arg to @$@ needs to be a function. does haskell know to treat
+-- @e@ as the second arg based on type? no, that's not it.
+-- oh, it must be related to being an infix operator, given the following types:
+
+-- Prelude> :t ($ 3)
+-- ($ 3) :: Num a => (a -> b) -> b
+-- Prelude> :t (($))
+-- (($)) :: (a -> b) -> a -> b
+-- Prelude> :t (($) 3)
+-- (($) 3) :: Num (a -> b) => a -> b
+-- Prelude> :t (3 $)
+-- (3 $) :: Num (a -> b) => a -> b
+
+-- I wonder how I would write this implementation if @$@ was not infix...
+
+-- btw, I really wanted to be able to pattern match on the higher-kinded type like:
+-- @apply e (T f) = T (f e)@
+-- but I couldn't figure out how to do it / if it is possible
+-- (is it possible?)
```

Comment:
ah, maybe it's not possible bc I wouldn't know how to call the constructor of the type? but how many ways can a kind parameterized by a single type be constructed? I guess it depends on the semantics of the kind? maybe there exists meaningful generalizations of this though?